README.md
# Xburn Laser Gcode Tool

TkinterUI
UI for xburn components (Incomplete)

### Prerequisites
python 2.7.x, PIL, numpy, Tkinter(tkinter)


---------------------------------------------------------

Python Command Line Tool
Create gcode files for a laser mounted to a shapeoko3 using stock firmware

-Skip's White Areas
-Burns in both directions, reversing every other line

<p align="center">
  <a href="https://www.youtube.com//watch?v=BWzQP15pxRQ"
  target="_blank"><img src="https://i.ytimg.com/vi_webp/BWzQP15pxRQ/maxresdefault.webp"
  width="350"/>
  </a>
</p>


-TODO - FIX Shades Option. Howto

### Prerequisites

python, pil?, numpy?

### Installing

```
git clone https://github.com/Emerica/xburn.git
cd xburn
pip install -r requirements.txt
chmod +x cli.py
./cli.py -h
usage: cli.py [-h] [-v] [-pa] [-s SHADES] [-wv WHITEVALUE] [-xd XDENSITY]
              [-yd YDENSITY] [-sr SKIPRATE] [-br BURNRATE] [-st STEPS]
              [-hp HIGHPOWER] [-lp LOWPOWER] [-on LASERON] [-off LASEROFF]
              [-mod MODIFIER] [-o OUTPUT] [-p] [-tp] [-d]
              file width

positional arguments:
  file                  image file name
  width                 Output width in MM (ish), FIX ME

optional arguments:
  -h, --help            show this help message and exit
  -v, --version         show program's version number and exit
  -pa, --palette        Color Palette, use with shades option, needs work.
  -s SHADES, --shades SHADES
                        Number of shades, default 16
  -wv WHITEVALUE, --whitevalue WHITEVALUE
                        White value, defaults to 255, anything larger than
                        this is skipped.
  -xd XDENSITY, --xdensity XDENSITY
                        Pixels per MM in X direction, default 3
  -yd YDENSITY, --ydensity YDENSITY
                        Pixels per MM Y direction, default 3
  -sr SKIPRATE, --skiprate SKIPRATE
                        Moving Feed Rate
  -br BURNRATE, --burnrate BURNRATE
                        Burning Feed Rate
  -st STEPS, --steps STEPS
                        Laser PWM Steps
  -hp HIGHPOWER, --highpower HIGHPOWER
                        Laser Max Power PWM VAlUE
  -lp LOWPOWER, --lowpower LOWPOWER
                        Laser Min Power PWM VAlUE
  -on LASERON, --laseron LASERON
                        Laser ON Gcode Command default: M3
  -off LASEROFF, --laseroff LASEROFF
                        Laser Off Gcode Command default: M5
  -mod MODIFIER, --modifier MODIFIER
                        Laser Power Modifier, defaults to Spindle Speed (S)
  -o OUTPUT, --output OUTPUT
                        Outfile name prefix
  -p, --preview         Preview burn output, red is skipped over.
  -tp, --testpattern    Create a test pattern. Use ./cli.py test 100 -tp -p -o
                        testfile
  -d, --debug           Turns on Debugging
```

Output ~100mm wide image with preview enabled


```
./cli.py filename.jpg 100 -o filename -p
cat filename.gcode
```


Create a gradient test pattern for tuning power levels

```
./cli.py test 100 -tp -o testpattern -p [non default settings here]
cat testpattern.gcode

```


CHANGELOG.md
# ChangeLog

## [0.8] - 2017-04-02
### Added
- UI added using Tkinter (still in the works)

### Changed
- Removed force saving processed image in cli.py


ui.py
from __future__ import division
from __future__ import print_function

import math
import os
import subprocess
import tkFileDialog
import ttk
from Tkinter import *
from itertools import groupby

import PIL.Image
import PIL.ImageTk
import numpy

# instance for Tkinter
root = Tk()


class xBurnObject(object):
    filename = StringVar()
    width = IntVar()
    shades = IntVar()
    white_value = IntVar()
    x_density = DoubleVar()
    y_density = DoubleVar()
    palette_on = BooleanVar()
    skip_rate = IntVar()
    burn_rate = IntVar()
    steps = IntVar()
    max_power = IntVar()
    min_power = IntVar()
    laser_on = StringVar()
    laser_off = StringVar()
    laser_mod = StringVar()
    out_file_prefix = StringVar()
    preview_on = BooleanVar()
    debugging_on = BooleanVar()

    def __init__(self, filename, width, shades, white_values, x_density, y_density, palette_on, skip_rate, burn_rate,
                 steps,
                 max_power, min_power, laser_on, laser_off, laser_mod, out_file_prefix, preview_on, debugging_on):
        self.filename.set(filename)
        self.width.set(width)
        self.shades.set(shades)
        self.white_value.set(white_values)
        self.x_density.set(x_density)
        self.y_density.set(y_density)
        self.palette_on.set(palette_on)
        self.skip_rate.set(skip_rate)
        self.burn_rate.set(burn_rate)
        self.steps.set(steps)
        self.max_power.set(max_power)
        self.min_power.set(min_power)
        self.laser_on.set(laser_on)
        self.laser_off.set(laser_off)
        self.laser_mod.set(laser_mod)
        self.out_file_prefix.set(out_file_prefix)
        self.preview_on.set(preview_on)
        self.debugging_on.set(debugging_on)


class XBurnView:
    label1 = None
    img_path = ""
    xburn = xBurnObject(img_path, 50, 16, 255, 2.0, 2.0, True, 3000, 800, 255, 12000, 0, "M5", "M3", "S", "workfile",
                        True, True)

    def __init__(self, master):
        self.master = master
        # root.minsize(width=1024, height=768)
        root.title("xBurn")

        content = ttk.Frame(master, padding=(3, 3, 12, 12))
        content.grid(column=0, sticky=(N, S, E, W))

        # Frame 1 stuff
        frame1 = Frame(content, bg="grey", borderwidth=5, relief="sunken", width=512, height=768)
        frame1.grid(column=0, row=0, columnspan=3, rowspan=2, sticky=(N, S, E, W))

        # Image to burn vars
        iFrame = Frame(frame1, bg="yellow", borderwidth=5, relief="sunken", width=512)
        iFrame.grid(column=0, row=0, columnspan=3, rowspan=2, sticky=(N, S, E, W))

        # Label(frame1, text="Number of shades", underline=True).grid(column=0, row=1, sticky=W, pady=5, padx=5)
        desired_width = Scale(iFrame, from_=0, to=500, length=300, label="Width to resize image in mm, defaults to 50",
                              orient=HORIZONTAL, variable=self.xburn.width)
        desired_width.grid(column=0, row=0, sticky=W, pady=5, padx=5)

        # Label(frame1, text="Number of shades", underline=True).grid(column=0, row=1, sticky=W, pady=5, padx=5)
        shades_num = Scale(iFrame, from_=1, to=257, length=300, label="Number of shades, defaults to 16",
                           orient=HORIZONTAL, variable=self.xburn.shades)
        shades_num.grid(column=0, row=1, sticky=W, pady=5, padx=5)

        # Label(frame1, text="White value", underline=True).grid(column=0, row=3, sticky=W, pady=5, padx=5)
        white_value = Scale(iFrame, from_=0, to=255, length=256, label="White value, defaults to 255",
                            orient=HORIZONTAL, variable=self.xburn.white_value)
        white_value.grid(column=1, row=1, sticky=E, pady=5, padx=5)

        # Label(frame1, text="x Density", underline=True).grid(column=0, row=5, sticky=W, pady=5, padx=5)
        x_density = Scale(iFrame, from_=0, to=10, length=256, resolution=0.1,
                          label="Pixels per mm in x direction, default 2.0", orient=HORIZONTAL,
                          variable=self.xburn.x_density)
        x_density.grid(column=0, row=2, sticky=W, pady=5, padx=5)

        # Label(frame1, text="y Density", underline=True).grid(column=0, row=7, sticky=W, pady=5, padx=5)
        y_density = Scale(iFrame, from_=0, to=10, length=256, resolution=0.1,
                          label="Pixels per mm in y direction, default 2.0", orient=HORIZONTAL,
                          variable=self.xburn.y_density)
        y_density.grid(column=1, row=2, sticky=E, pady=5, padx=5)

        palette = Checkbutton(iFrame, text="Palette", variable=self.xburn.palette_on)
        palette.grid(column=0, row=3, sticky=W, pady=5, padx=5)

        # Laser vars frame
        lFrame = Frame(frame1, bg="red", borderwidth=5, relief="sunken", width=512)
        lFrame.grid(column=0, row=2, columnspan=3, rowspan=2, sticky=(N, S, E, W))

        # Label(frame1, text="Skip rate", underline=True).grid(column=0, row=9, sticky=W, pady=5, padx=5)
        skip_rate = Scale(lFrame, from_=0, to=3000, length=256, label="Moving Feed Rate, defaults to 3000",
                          orient=HORIZONTAL, variable=self.xburn.skip_rate)
        skip_rate.grid(column=0, row=0, sticky=W, pady=5, padx=5)

        # Label(frame1, text="Burn rate", underline=True).grid(column=0, row=11, sticky=W, pady=5, padx=5)
        burn_rate = Scale(lFrame, from_=0, to=1000, length=256, label="Burning Feed Rate, defaults to 800",
                          orient=HORIZONTAL, variable=self.xburn.burn_rate)
        burn_rate.grid(column=1, row=0, sticky=W, pady=5, padx=5)

        # Label(frame1, text="Steps", underline=True).grid(column=0, row=13, sticky=W, pady=5, padx=5)
        steps = Scale(lFrame, from_=0, to=500, length=256, label="Laser PWM Steps, defaults to 255",
                      orient=HORIZONTAL, variable=self.xburn.steps)
        steps.grid(column=2, row=0, sticky=W, pady=5, padx=5)

        # Label(frame1, text="Max power", underline=True).grid(column=0, row=15, sticky=W, pady=5, padx=5)
        max_power = Scale(lFrame, from_=0, to=15000, length=300, label="Laser Max Power PWM VAlUE, defaults to 12000",
                          orient=HORIZONTAL, variable=self.xburn.max_power)
        max_power.grid(column=0, row=1, sticky=W, pady=5, padx=5)

        # Label(frame1, text="Min power", underline=True).grid(column=0, row=17, sticky=W, pady=5, padx=5)
        min_power = Scale(lFrame, from_=0, to=15000, length=300, label="Laser Min Power PWM VAlUE, defaults to 0",
                          orient=HORIZONTAL, variable=self.xburn.min_power)
        min_power.grid(column=1, row=1, sticky=E, pady=5, padx=5)

        Label(lFrame, text="Laser on command", underline=True).grid(column=0, row=2, sticky=W, pady=5, padx=5)
        laser_on = Entry(lFrame, width=20, textvariable=self.xburn.laser_on)
        laser_on.grid(column=0, row=3, sticky=W, pady=5, padx=5)

        Label(lFrame, text="Laser off command", underline=True).grid(column=1, row=2, sticky=W, pady=5, padx=5)
        laser_off = Entry(lFrame, width=20, textvariable=self.xburn.laser_off)
        laser_off.grid(column=1, row=3, sticky=W, pady=5, padx=5)

        Label(lFrame, text="Laser power modifier", underline=True).grid(column=2, row=2, sticky=W, pady=5, padx=5)
        laser_power_mod = Entry(lFrame, width=20, textvariable=self.xburn.laser_mod)
        laser_power_mod.grid(column=2, row=3, sticky=W, pady=5, padx=5)

        # Misc stuff
        mFrame = Frame(frame1, bg="blue", borderwidth=5, relief="sunken", width=512)
        mFrame.grid(column=0, row=4, columnspan=3, rowspan=2, sticky=(N, S, E, W))

        Label(mFrame, text="Outfile name", underline=True).grid(column=0, row=0, sticky=W, pady=5, padx=5)
        outfile_name = Entry(mFrame, width=50, textvariable=self.xburn.out_file_prefix)
        outfile_name.grid(column=0, row=1, sticky=W, pady=5, padx=5)

        check1 = Checkbutton(mFrame, text="Preview", variable=self.xburn.preview_on)
        check1.grid(column=2, row=0, sticky=W, pady=5, padx=5)

        check2 = Checkbutton(mFrame, text="Debug", variable=self.xburn.debugging_on)
        check2.grid(column=2, row=1, pady=5, padx=5)

        # Frame 2 stuff
        frame2 = Frame(content, bg="green", borderwidth=5, relief="sunken")
        frame2.grid(column=3, row=0, columnspan=3, rowspan=2, sticky=(N, S, E, W))

        global label1
        label1 = Label(frame2, text="Upload image")
        label1.grid(row=0, column=0, sticky=N, pady=5, padx=5)
        # self.get_resize_update_img("mysterion.gif")

        button1 = Button(frame2, text="Upload image", width=20, command=self.upload_image)
        button1.grid(row=1, column=0, sticky=S, pady=5, padx=5)

        button2 = Button(frame2, text="Gradient test", width=20, command=self.gradient_test)
        button2.grid(row=2, column=0, sticky=S, pady=5, padx=5)

        button3 = Button(frame2, text="Convert image", width=20, command=self.convert_all)
        button3.grid(row=3, column=0, sticky=S, pady=5, padx=5)

        button4 = Button(frame2, text="Generate gcode", width=20, command=self.generate_gcode)
        button4.grid(row=4, column=0, sticky=S, pady=5, padx=5)

    def upload_image(self):
        global img_path
        img_path = tkFileDialog.askopenfilename()

        self.xburn.filename.set(img_path)

        self.get_resize_update_img(img_path)

    def generate_gcode(self):
        print('Width ' + str(self.xburn.width.get()))
        print('Shades ' + str(self.xburn.shades.get()))
        print('White value ' + str(self.xburn.white_value.get()))
        print('x density ' + str(self.xburn.x_density.get()))
        print('y density ' + str(self.xburn.y_density.get()))
        print('Density(max(x,y) ' + str(max(self.xburn.x_density.get(), self.xburn.y_density.get())))
        print('Skip rate ' + str(self.xburn.skip_rate.get()))
        print('Burn rate ' + str(self.xburn.burn_rate.get()))
        print('Steps ' + str(self.xburn.steps.get()))
        print('Max power ' + str(self.xburn.max_power.get()))
        print('Min power ' + str(self.xburn.min_power.get()))
        print('Laser on mapping ' + str(self.xburn.laser_on.get()))
        print('Laser off mapping ' + str(self.xburn.laser_off.get()))
        print('Laser mod ' + str(self.xburn.laser_mod.get()))
        print('Outfile prefix ' + str(self.xburn.out_file_prefix.get()))
        print('Preview ON ' + str(self.xburn.preview_on.get()))
        print('Debug ON ' + str(self.xburn.debugging_on.get()))

        # TODO replace with the proper calls
        # start cli process
        path = os.path.join(os.getcwd(), 'cli.py ')
        cmd = path + (self.xburn.filename.get() + ' ') \
              + (str(self.xburn.width.get()) + ' ') \
              + ('-s ' + str(self.xburn.shades.get()) + ' ') \
              + ('-wv ' + str(self.xburn.white_value.get()) + ' ') \
              + ('-de ' + str(max(self.xburn.x_density.get(), self.xburn.y_density.get())) + ' ') \
              + ('-sr ' + str(self.xburn.skip_rate.get()) + ' ') \
              + ('-br ' + str(self.xburn.burn_rate.get()) + ' ') \
              + ('-st ' + str(self.xburn.steps.get()) + ' ') \
              + ('-hp ' + str(self.xburn.max_power.get()) + ' ') \
              + ('-lp ' + str(self.xburn.min_power.get()) + ' ') \
              + ('-on ' + str(self.xburn.laser_on.get()) + ' ') \
              + ('-off ' + str(self.xburn.laser_off.get()) + ' ') \
              + ('-mod ' + str(self.xburn.laser_mod.get()) + ' ') \
              + ('-o ' + str(self.xburn.out_file_prefix.get()) + ' ') \
              + ('-p ' if self.xburn.preview_on.get() else '') \
              + ('-d ' if self.xburn.debugging_on.get() else '')

        print(cmd)

        subprocess.Popen(cmd, shell=True)

    def get_resize_update_img(self, path):
        global img_path
        img_path = path

        base_width = 400
        img = PIL.Image.open(img_path)
        [img_width, img_height] = img.size

        width_ratio = (base_width / float(img_width))
        new_height = int((float(img_height) * float(width_ratio)))

        img = img.resize((base_width, new_height), PIL.Image.ANTIALIAS)

        img2 = PIL.ImageTk.PhotoImage(img)
        label1.configure(image=img2)
        label1.image = img2

        return img

    def resize_update_img(self, image):
        base_width = 400
        img = image
        [img_width, img_height] = img.size

        width_ratio = (base_width / float(img_width))
        new_height = int((float(img_height) * float(width_ratio)))

        img = img.resize((base_width, new_height), PIL.Image.ANTIALIAS)

        img2 = PIL.ImageTk.PhotoImage(img)
        label1.configure(image=img2)
        label1.image = img2

        return img

    def convert_image(self):
        img = PIL.Image.open(img_path)
        [img_width, img_height] = img.size

        img = img.convert("L")

        height_ratio = self.xburn.width.get() / (img_width / img_height)
        resize_co = (int(math.floor(self.xburn.x_density.get() * self.xburn.width.get())),
                     int(math.floor(self.xburn.y_density.get() * height_ratio)))

        if self.xburn.palette_on.get():
            if self.xburn.shades.get() < 3:
                palette = [0, 0, 0, 255, 255, 255, ] + [255, ] * 254 * 3
            else:
                palette = [0, 0, 0, ]
                steps = int(math.floor(256 / (self.xburn.shades.get() - 1)))
                for c in range(self.xburn.shades.get() - 2):
                    m = c + 1
                    palette = palette + [steps * m, steps * m, steps * m]
                palette = palette + [255, ] * (256 - self.xburn.shades.get()) * 3

            pimage = PIL.Image.new("P", (1, 1), 0)
            pimage.putpalette(palette)

            img = img.quantize(colors=self.xburn.shades.get(), palette=pimage)

        img = img.resize(resize_co, PIL.Image.LANCZOS) \
            .transpose(PIL.Image.ROTATE_180) \
            .transpose(PIL.Image.FLIP_LEFT_RIGHT)

        # self.resize_update_img(img)

        return img

    def get_converted_img_array(self, img):
        return numpy.array(img)

    # creates a 255x20 black to white gradient for testing settings
    def gradient_test(self):
        test = PIL.Image.new('RGB', (255, 20), "black")  # create a new black image
        pixels = test.load()  # create the pixel map
        for i in range(test.size[0]):  # for every pixel:
            for j in range(test.size[1]):
                pixels[i, j] = (i, i, i)  # set the colour accordingly

        self.resize_update_img(test)

    def convert_all(self):
        img = self.convert_image()
        arr = self.get_converted_img_array(img)

        prv = PIL.Image.new('RGB', (len(arr[0]), len(arr)), "red")
        pixels = prv.load()  # create the pixel map

        # Y position
        yp = 0

        for y in arr:
            # If we have an even number for a y axis line
            if yp % 2 != 0:
                # Direction is reversed, set xp to the end
                xp = len(y) - 1
                # Revese the values of y
                y = list(reversed(y))
                rev = True
            else:
                xp = 0
                rev = False
            # Group pixels by value into a new list
            for i, j in groupby(y):
                # items in the list
                items = list(j)
                # Number of items
                size = len(items)
                # grey Value in this chunk of the line
                value = items[0]
                # Make sure this group isn't above the whitevalue
                if value < self.xburn.white_value.get():

                    pvx = len(items) - 1 if rev else 0
                    for item in items:
                        pix = xp - pvx if rev else xp + pvx
                        pixels[pix, yp] = (item, item, item)
                        pvx = pvx - 1 if rev else pvx + 1

                # Increment position
                xp = xp - size if rev else xp + size

            yp = yp + 1

        prv.transpose(PIL.Image.ROTATE_180).transpose(PIL.Image.FLIP_LEFT_RIGHT)

        self.resize_update_img(prv)


xburn = XBurnView(root)
root.mainloop()


cli.py
#!/usr/bin/env python
from __future__ import division

import argparse
import math

import numpy

try:
    import Image
except ImportError:
    from PIL import Image

from itertools import groupby

version = "0.8"


def loadArray(arr):
    im = Image.fromarray(arr)


def loadImage(file):
    # Open Source
    img = Image.open(file)
    # aspect
    height = args.width / (img.size[0] / img.size[1])
    # This resize is close
    # Rotated and flipped to look proper
    resize = (
        int(math.floor(args.density * args.width)),
        int(math.floor(args.density * height))
    )
    # TODO: make quantize work with a palette, better.
    if args.palette:
        if args.shades < 3:
            palette = [
                          0, 0, 0,
                          255, 255, 255,
                      ] + [255, ] * 254 * 3
        else:
            palette = [0, 0, 0, ]
            steps = int(math.floor(256 / (args.shades - 1)))
            for c in range(args.shades - 2):
                m = c + 1
                palette = palette + [steps * m, steps * m, steps * m]
            palette = palette + [255, ] * (256 - args.shades) * 3
        pimage = Image.new("P", (1, 1), 0)
        pimage.putpalette(palette)
        if args.debug:
            print "Number of shades:" + str(args.shades)
            print "Number of steps:" + str(steps)
            print palette
        # .convert("L", palette=pimage) \
        img = img.convert("L") \
            .quantize(colors=args.shades, palette=pimage) \
            .resize(resize, Image.LANCZOS) \
            .transpose(Image.ROTATE_180) \
            .transpose(Image.FLIP_LEFT_RIGHT) \
            # .save(args.output+".gcode.jpg")
        # .show()
    else:
        if args.debug:
            print "No palette supplied"
        img = img.convert("L") \
            .resize(resize, Image.LANCZOS) \
            .transpose(Image.ROTATE_180) \
            .transpose(Image.FLIP_LEFT_RIGHT) \
            # .save(args.output+".gcode.jpg")
        # .show()
    # Save temp, TODO: do without writing file? tired.
    # img = Image.open(args.output+".gcode.jpg")
    # img.show()
    imgarr = numpy.array(img)
    return imgarr


def laserOff():
    appendGcode(args.laseroff)


def laserOn(power):
    appendGcode(args.modifier + str(power) + " " + args.laseron)


# creates a 255x20 black to white gradient for testing settings
def gradientTest():
    test = Image.new('RGB', (255, 20), "black")  # create a new black image
    pixels = test.load()  # create the pixel map
    for i in range(test.size[0]):  # for every pixel:
        for j in range(test.size[1]):
            pixels[i, j] = (i, i, i)  # set the colour accordingly
    test.save("gradient_testpatten.jpg")


def appendGcode(line):
    global lines
    # Remove any duplicate commands...
    # optimizing where possible.
    numlines = len(lines)
    prefix = line[:2]
    if len(lines) > 0:
        if line == lines[numlines - 1]:
            pass
            # Duplicate gcode found. OMIT
            # if args.debug:
            #    print "OMIT DUPLICATE:"  + line
        elif prefix == lines[numlines - 1][:2] and "Y" not in lines[numlines - 1]:
            lines[numlines - 1] = line
            # if args.debug:
            #    print "OVERWRITE:"  + line
        elif args.laseron in line and args.laseroff in lines[numlines - 1]:
            lines[numlines - 1] = line
            # if args.debug:
            #    print "Keep laser on:"  + line
        else:
            lines.append(line)
    else:
        lines.append(line)


# Set laser mode for grbl 1.1
# Appears this should work? in practice?
def laserMode(status):
    if status == 1:
        appendGcode("$38=1")
    else:
        appendGcode("$38=0")


# TODO: config profiles so you don't have to mess around.....
# SERIOUSLY! ^^^^^^^
parser = argparse.ArgumentParser()
parser.add_argument('file', help='image file name')
parser.add_argument('width', type=int, help='Output width in MM (ish), FIX ME')

# Versioning
parser.add_argument('-v', '--version', action='version', version=version)
# TODO: arguments and profiles.....
# parser.add_argument('-s', '--size',  type=float, help='pixelsize')
parser.add_argument('-pa', '--palette', action='store_true',
                    help='Color Palette, use with shades option, needs work.')
parser.add_argument('-s', '--shades', type=int, default=16,
                    help='Number of shades, default 16')
parser.add_argument('-wv', '--whitevalue', type=int, default=255,
                    help='White value, defaults to 255, anything larger than this is skipped.')
parser.add_argument('-de', '--density', type=float, default=2.0,
                    help='Pixels per MM, default 2.0')
parser.add_argument('-sr', '--skiprate', type=int, default=3000,
                    help='Moving Feed Rate')
parser.add_argument('-br', '--burnrate', type=int, default=800,
                    help='Burning Feed Rate')
parser.add_argument('-st', '--steps', type=int, default=255,
                    help='Laser PWM Steps')
parser.add_argument('-hp', '--highpower', type=int, default=12000,
                    help='Laser Max Power PWM VAlUE')
parser.add_argument('-lp', '--lowpower', type=int, default=0,
                    help='Laser Min Power PWM VAlUE')
parser.add_argument('-on', '--laseron', default="M3",
                    help='Laser ON Gcode Command default: M3')
parser.add_argument('-off', '--laseroff', default="M5",
                    help='Laser Off Gcode Command default: M5')
parser.add_argument('-mod', '--modifier', default="S",
                    help='Laser Power Modifier, defaults to Spindle Speed (S)')
parser.add_argument('-o', '--output', default="workfile",
                    help='Outfile name prefix')
parser.add_argument('-p', '--preview', action='store_true',
                    help='Preview burn output, red is skipped over.')
parser.add_argument('-gr', '--grblver', type=float, default=0.9,
                    help='Default Grbl version is 0.9')
parser.add_argument('-tp', '--testpattern', action='store_true',
                    help='Create a test pattern. Use ./cli.py test 100 -tp -p -o testfile')
parser.add_argument('-d', '--debug', action='store_true',
                    help='Turns on Debugging')

# Check the arguments
args = parser.parse_args()

# Make sure at least one option is chosen
if not (args.file or args.testpattern):
    # Print help
    parser.print_help()
    # Exit out with no action message
    parser.error('No action requested')

if args.lowpower == 0:
    args.lowpower = math.ceil(args.highpower / args.steps)
    if args.debug:
        print "Low power set to: " + str(args.lowpower)

if args.testpattern:
    gradientTest()
    args.file = "gradient_testpatten.jpg"

# Do all the things
if args.file:
    # Load a image file to array
    arr = loadImage(args.file)
    scaley = 1 / args.density
    scalex = 1 / args.density
    # Create a list to store the output gcode lines
    lines = []
    appendGcode(";Xburn: " + str(args))
    # Y position
    yp = 0
    # if the grbl version allows lasermode, enable it
    if args.grblver > 0.9:
        laserMode(1)
    # Turn the laser off
    laserOff()
    if args.preview:
        prv = Image.new('RGB', (len(arr[0]), len(arr)), "red")
        pixels = prv.load()  # create the pixel map
    # Work in MM
    appendGcode("G21")
    # Loop over the list
    for y in arr:
        # If we have an even number for a y axis line
        if yp % 2 != 0:
            # Direction is reversed, set xp to the end
            xp = len(y) - 1
            # Revese the values of y
            y = list(reversed(y))
            rev = True
        else:
            xp = 0
            rev = False
        start = False
        # reset the lastxp position
        lastxp = 0
        # Group pixels by value into a new list
        for i, j in groupby(y):
            # items in the list
            items = list(j)
            # Number of items
            size = len(items)
            # grey Value in this chunk of the line
            value = items[0]
            # Make sure this group isn't above the whitevalue
            if value < args.whitevalue:
                if start == False:
                    # Go the start of this line, the first place with a value
                    appendGcode("G0 X" + str(round(xp * scaley, 3)) + " Y" +
                                str(round(yp * scaley, 3)) + " F" + str(args.skiprate))
                    start = True
                else:
                    appendGcode("G0 X" + str(round(xp * scalex, 3)) +
                                " F" + str(args.skiprate))
                # Create the preview
                if args.preview:
                    pvx = len(items) - 1 if rev else 0
                    for item in items:
                        pix = xp - pvx if rev else xp + pvx
                        pixels[pix, yp] = (item, item, item)
                        pvx = pvx - 1 if rev else pvx + 1
                # Turn on the laser
                step = (args.highpower - args.lowpower) / args.steps
                laserOn(math.ceil((args.steps - value) * step))
                # Burn the segment
                goto = xp - size if rev else xp + size
                appendGcode("G1 X" + str(round((goto) * scalex, 3)) +
                            " F" + str(args.burnrate))
                laserOff()
            else:
                # skip the white
                if xp > 0 and xp < len(y) and lastxp == xp:
                    goto = xp - size if rev else xp + size
                    appendGcode("G0 X" + str(round((goto) * scalex, 3)) +
                                " F" + str(args.skiprate))
            # track x position
            lastxp = xp
            # Increment position
            xp = xp - size if rev else xp + size
        yp = yp + 1
        # Turn the laser off
        laserOff()
    # Go to zero
    appendGcode("G00 X0 Y0 F" + str(args.skiprate))
    # if the grbl version allows lasermode, disable it
    if args.grblver > 0.9:
        laserMode(0)
    # Show a preview if enabled
    if args.preview:
        prv.transpose(Image.ROTATE_180).transpose(Image.FLIP_LEFT_RIGHT).show()
    # Output the gcode file
    f = open(args.output + '.gcode', 'w')
    f.write("\n".join(lines))
